src/result-async.ts

```diff
@@ -0,0 +1,123 @@
+import { Result, Ok, Err } from './'
+import { isPromise } from './'
+
+export class ResultAsync<T, E> {
+  private _promise: Promise<Result<T, E>>
+
+  constructor(res: Promise<Result<T, E>>) {
+    this._promise = res
+  }
+
+  static fromPromise<T, E>(promise: Promise<T>): ResultAsync<T, E> {
+    return new ResultAsync(promise.then((value: T) => new Ok<T, E>(value)))
+  }
+
+  map<A>(_f: (t: T) => A | Promise<A>): ResultAsync<A, E> {
+    const newPromise: Promise<Result<A, E>> = this._promise.then((res: Result<T, E>) => {
+      if (res.isErr()) {
+        return new Err<A, E>(res.error)
+      }
+
+      const newValue = _f(res.value)
+
+      if (isPromise(newValue)) {
+        return newValue.then(value => new Ok<A, E>(value))
+      }
+
+      return new Ok<A, E>(newValue)
+    })
+
+    return new ResultAsync(newPromise)
+  }
+
+  mapErr<U>(_f: (e: E) => U | Promise<U>): ResultAsync<T, U> {
+    const newPromise: Promise<Result<T, U>> = this._promise.then((res: Result<T, E>) => {
+      if (res.isOk()) {
+        return new Ok<T, U>(res.value)
+      }
+
+      const newError = _f(res.error)
+
+      if (isPromise(newError)) {
+        return newError.then(error => new Err<T, U>(error))
+      }
+
+      return new Err<T, U>(newError)
+    })
+
+    return new ResultAsync(newPromise)
+  }
+
+  andThen<U>(_f: (t: T) => Result<U, E> | ResultAsync<U, E>): ResultAsync<U, E> {
+    const newPromise = this._promise.then((res: Result<T, E>) => {
+      if (res.isErr()) {
+        return new Err<U, E>(res.error)
+      }
+
+      const newValue = _f(res.value)
+
+      if (newValue instanceof ResultAsync) {
+        return newValue._promise
+      }
+
+      return newValue
+    })
+
+    return new ResultAsync(newPromise)
+  }
+
+  match = <A>(ok: (t: T) => A, _err: (e: E) => A): Promise<A> => {
+    return this._promise.then(res => res.match(ok, _err))
+  }
+
+  // Makes ResultAsync awaitable
+  then(_successCallback: (res: Result<T, E>) => void) {
+    this._promise.then(_successCallback)
+  }
+}
+
+export const okAsync = <T, E>(value: T): ResultAsync<T, E> =>
+  new ResultAsync(Promise.resolve(new Ok<T, E>(value)))
+
+export const errAsync = <T, E>(err: E): ResultAsync<T, E> =>
+  new ResultAsync(Promise.resolve(new Err<T, E>(err)))
```

Comment:
What's the use case for having `okAsync` and `errAsync`? It seems to me that if I want to convert a normal result into a async result, I would call `.asyncMap`, no?

I would prefer to keep the surface area of code as small as possible. This is where I'm coming from when I ask this question, and If there's no good reason right now to keep `okAsync` and `errAsync`, then I would ask that it be removed.

Reply:
Here's an example use-case:

```typescript
const createUser = (login: string): ResultAsync<boolean, Error> => {
  if (login.length < 3) {
    return errAsync(new Error('Username is too short'))
  }

  return ResultAsync.fromPromise(
    insertUserIntoDb(login),
    () => new Error('Failed to insert into database'),
  )
}
```

=> A method that should return a `ResultAsync` and also performs synchronous operations. It's the async equivalent of `Result`s `ok` and `err` handlers.
It avoids writing `return new ResultAsync(Promise.resolve(new Err<T, E>(err)))`...